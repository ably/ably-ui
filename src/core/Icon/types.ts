import { computedIcons } from "./computed-icons";

export const iconNames = {
  gui: computedIcons.gui,
  display: computedIcons.display,
  social: computedIcons.social,
  other: computedIcons.other,
  tech: computedIcons.tech,
  product: computedIcons.product,
};

export type IconName =
  | (typeof iconNames.gui)[number]
  | (typeof iconNames.display)[number]
  | (typeof iconNames.social)[number]
  | (typeof iconNames.other)[number]
  | (typeof iconNames.tech)[number]
  | (typeof iconNames.product)[number];
