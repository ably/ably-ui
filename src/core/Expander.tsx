import React, { PropsWithChildren, useEffect, useRef, useState } from "react";
import throttle from "lodash.throttle";

type ExpanderProps = {
  heightThreshold?: number;
  className?: string;
  fadeClassName?: string;
  controlsClassName?: string;
  controlsOpenedLabel?: string;
  controlsClosedLabel?: string;
};

const Expander = ({
  heightThreshold = 200,
  className,
  fadeClassName,
  controlsClassName,
  controlsOpenedLabel,
  controlsClosedLabel,
  children,
}: PropsWithChildren<ExpanderProps>) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const [showControls, setShowControls] = useState(false);
  const [contentHeight, setContentHeight] = useState<number>(heightThreshold);
  const [height, setHeight] = useState<number | "auto">(heightThreshold);
  const [expanded, setExpanded] = useState(false);

  useEffect(() => {
    if (innerRef.current) {
      setContentHeight(innerRef.current.clientHeight);
    }

    if (contentHeight < heightThreshold) {
      setHeight("auto");
    } else if (expanded) {
      setHeight(contentHeight);
    } else {
      setHeight(heightThreshold);
    }

    setShowControls(contentHeight >= heightThreshold);
  }, [contentHeight, heightThreshold, expanded]);

  useEffect(() => {
    const onResize = throttle(() => {
      if (innerRef.current) {
        setContentHeight(innerRef.current.clientHeight);
      }
    }, 250);

    window.addEventListener("resize", onResize);
    return () => {
      window.removeEventListener("resize", onResize);
    };
  }, []);

  return (
    <>
      <div
        style={{ height }}
        data-testid="expander-container"
        className={`overflow-hidden transition-all relative ${className ?? ""}`}
      >
        {showControls && !expanded && (
          <div
            className={`h-64 w-full bg-gradient-to-t from-white to-transparent absolute bottom-0 left-0 right-0 ${
              fadeClassName ?? ""
            }`}
          ></div>
        )}
        <div ref={innerRef}>{children}</div>
      </div>
      {showControls && (
        <div
          onClick={() => setExpanded(!expanded)}
          onKeyDown={(e) => e.key === "Enter" && setExpanded(!expanded)}
          tabIndex={0}
          data-testid="expander-controls"
          className={`mt-16 cursor-pointer font-bold text-gui-blue-default-light hover:text-gui-blue-hover-light ${controlsClassName ?? ""}`}
        >
          {expanded
            ? (controlsOpenedLabel ?? "View less -")
            : (controlsClosedLabel ?? "View all +")}
        </div>
      )}
    </>
  );
};

export default Expander;
