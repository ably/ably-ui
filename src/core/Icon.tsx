import React, { CSSProperties, useCallback, useId, useMemo } from "react";

import {
  defaultIconSecondaryColor,
  getHeroicon,
  setUniqueIds,
  toPascalCase,
} from "./Icon/utils";
import * as IconComponents from "./Icon/components";
import { IconName, IconSize } from "./Icon/types";
import { ColorClass, ColorThemeSet } from "./styles/colors/types";
import { convertTailwindClassToVar } from "./styles/colors/utils";
import cn from "./utils/cn";

export type IconProps = {
  name: IconName;
  size?: IconSize;
  color?: ColorClass | ColorThemeSet;
  secondaryColor?: ColorClass | ColorThemeSet;
  additionalCSS?: string;
  variant?: "outline" | "solid" | "micro" | "mini";
};

// Helper function to get a heroicon component from the static map
const getHeroiconComponent = (
  iconName: string,
): React.ComponentType<React.SVGProps<SVGSVGElement>> | null => {
  // Parse the icon name to extract heroicon name and variant
  // Expected format: icon-gui-{heroicon-name}-{variant}
  const match = iconName.match(/^icon-gui-(.+)-(outline|solid|mini|micro)$/);
  if (!match) {
    return null;
  }

  const [, heroiconName, variant] = match;
  const component = getHeroicon(heroiconName, variant);
  return component;
};

const Icon = ({
  name,
  size = "0.75rem",
  color,
  secondaryColor,
  additionalCSS = "",
  variant,
  ...additionalAttributes
}: IconProps) => {
  const uniqueId = useId();

  const [lightSecondaryColor, darkSecondaryColor] = (
    secondaryColor ?? ""
  ).split(" dark:") as [ColorClass, ColorClass | undefined];

  const iconName =
    variant && !name.endsWith(`-${variant}`) ? `${name}-${variant}` : name;

  // Get the local icon component (memoized)
  const LocalIconComponent = useMemo(():
    | React.ComponentType<React.SVGProps<SVGSVGElement>>
    | undefined => {
    return IconComponents[
      toPascalCase(iconName) as keyof typeof IconComponents
    ];
  }, [iconName]);

  // Get the heroicon component only if no local component is found (memoized)
  const HeroiconComponent = useMemo(() => {
    return LocalIconComponent ? null : getHeroiconComponent(iconName);
  }, [LocalIconComponent, iconName]);

  const setUniqueIdsRef = useCallback(
    (el: SVGSVGElement | null) => {
      setUniqueIds(el, uniqueId);
    },
    [uniqueId],
  );

  const iconSvg = useCallback(
    (secondaryColor?: ColorClass, isDark?: boolean, isThemed?: boolean) => {
      if (LocalIconComponent) {
        let secondaryColorValue;
        if (secondaryColor) {
          secondaryColorValue = convertTailwindClassToVar(secondaryColor);
        } else if (defaultIconSecondaryColor(name)) {
          secondaryColorValue = convertTailwindClassToVar(
            defaultIconSecondaryColor(name),
          );
        }

        return (
          <LocalIconComponent
            className={cn(
              { [`${color}`]: color },
              { [`${additionalCSS}`]: additionalCSS },
              {
                "hidden dark:block": secondaryColor && !isDark && isThemed,
              },
              {
                "dark:hidden": secondaryColor && isDark && isThemed,
              },
            )}
            style={
              {
                width: size,
                height: size,
                ...(secondaryColorValue && {
                  "--ui-icon-secondary-color": secondaryColorValue,
                }),
              } as CSSProperties
            }
            {...additionalAttributes}
            ref={setUniqueIdsRef}
          />
        );
      }

      if (HeroiconComponent) {
        return (
          <HeroiconComponent
            className={cn(
              { [`${color}`]: color },
              { [`${additionalCSS}`]: additionalCSS },
            )}
            style={{
              width: size,
              height: size,
            }}
            {...additionalAttributes}
          />
        );
      }
    },
    [
      LocalIconComponent,
      HeroiconComponent,
      iconName,
      size,
      color,
      additionalCSS,
      lightSecondaryColor,
      darkSecondaryColor,
      name,
    ],
  );

  if (lightSecondaryColor && darkSecondaryColor) {
    return (
      <>
        {iconSvg(lightSecondaryColor, false, !!darkSecondaryColor)}
        {iconSvg(darkSecondaryColor, true, true)}
      </>
    );
  } else if (lightSecondaryColor) {
    return iconSvg(lightSecondaryColor, false, !!darkSecondaryColor);
  } else {
    return iconSvg();
  }
};

export default Icon;
