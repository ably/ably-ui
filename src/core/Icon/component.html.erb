<%= tag.svg class: "#{color} #{additional_css}", style: "width: #{size}; height: #{size}", **additional_attributes do %>
  <%= tag.use "xlink:href": "#sprite-#{name}" %>
<% end %>
