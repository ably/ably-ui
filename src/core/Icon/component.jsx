import React from "react";
import T from "prop-types";

const Icon = ({ name, size = "0.75rem", color = "", additionalCSS = "", ...additionalAttributes }) => {
  return (
    <svg className={`${color} ${additionalCSS}`} style={{ width: size, height: size }} {...additionalAttributes}>
      <use xlinkHref={`#sprite-${name}`} />
    </svg>
  );
};

Icon.propTypes = {
  name: T.string.isRequired,
  size: T.string,
  color: T.string,
  additionalCSS: T.string,
};

export default Icon;
