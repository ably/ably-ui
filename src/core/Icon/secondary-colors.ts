import { ColorClass } from "../styles/colors/types";
import { IconName } from "./types";

export const defaultIconSecondaryColor = (name: IconName) => {
  const colorLookup = {
    "icon-gui-check-circled-fill": "text-white",
  } as Record<IconName, ColorClass>;

  return colorLookup[name];
};
