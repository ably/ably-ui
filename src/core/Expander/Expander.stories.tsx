import React from "react";
import { userEvent, within, expect, waitFor } from "@storybook/test";
import Expander from "../Expander";
import Icon from "../Icon";

export default {
  title: "JS Components/Expander",
  component: Expander,
  tags: ["autodocs"],
  args: {
    height: 200,
  },
  argTypes: {
    height: {
      control: {
        type: "number",
      },
    },
  },
};

const longContentInner = (
  <div>
    <p>Ipsum</p>
    <ul className="mb-16 list-inside list-disc">
      <li>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</li>
      <li>Sed convallis ex pharetra, tristique tellus vel, rhoncus velit.</li>
      <li>Mauris molestie felis et scelerisque ullamcorper.</li>
      <li>Maecenas congue ligula ut commodo tristique.</li>
      <li>
        Pellentesque venenatis elit vitae urna condimentum, in mollis arcu
        venenatis.
      </li>
      <li>Donec nec turpis vel urna egestas fringilla.</li>
    </ul>
    <p>Ipsum</p>
    <ul className="mb-16 list-inside list-disc">
      <li>Mauris ut nibh vel metus cursus semper.</li>
      <li>Ut mattis tortor eu urna accumsan gravida.</li>
      <li>Nunc pellentesque neque at elit pretium tempor.</li>
      <li>Curabitur finibus magna vitae nunc varius fermentum.</li>
    </ul>
    <ul className="mb-16 list-inside list-disc">
      <li>Curabitur vehicula mi iaculis, luctus augue eu, venenatis quam.</li>
      <li>Praesent in eros efficitur, consequat ante eu, faucibus arcu.</li>
      <li>Nulla laoreet nibh a odio interdum, non molestie diam auctor.</li>
    </ul>
    <p>Ipsum</p>
    <ul className="mb-16 list-inside list-disc">
      <li>
        Praesent aliquam diam tincidunt, sollicitudin tortor eget, vulputate
        lacus.
      </li>
      <li>Quisque in mi sed ex vulputate varius in a leo.</li>
      <li>Etiam posuere dolor at tortor aliquam imperdiet.</li>
      <li>
        Maecenas quis neque consequat, ultricies est sit amet, congue est.
      </li>
      <li>Aenean a elit sed nibh pretium lacinia sed convallis sapien.</li>
    </ul>
    <p>Ipsum</p>
    <ul className="mb-16 list-inside list-disc">
      <li>
        Nulla malesuada libero id dolor aliquam, non sagittis mi scelerisque.
      </li>
      <li>
        Etiam tincidunt lacus eu diam laoreet consectetur sit amet non est.
      </li>
      <li>In porta arcu nec purus tincidunt vulputate.</li>
    </ul>
  </div>
);

export const LongContent = {
  render: () => <Expander>{longContentInner}</Expander>,
  parameters: {
    docs: {
      description: {
        story:
          "A larger amount of content that exceeds the height cut-off, controls shown.",
      },
    },
  },
};

export const ShortContent = {
  render: () => (
    <Expander>
      <div>
        <p>Ipsum</p>
        <ul className="mb-16 list-inside list-disc">
          <li>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</li>
          <li>
            Sed convallis ex pharetra, tristique tellus vel, rhoncus velit.
          </li>
          <li>Mauris molestie felis et scelerisque ullamcorper.</li>
          <li>Maecenas congue ligula ut commodo tristique.</li>
          <li>
            Pellentesque venenatis elit vitae urna condimentum, in mollis arcu
            venenatis.
          </li>
          <li>Donec nec turpis vel urna egestas fringilla.</li>
        </ul>
      </div>
    </Expander>
  ),
  parameters: {
    docs: {
      description: {
        story:
          "A smaller amount of content that doesn't exceed the height cut-off, therefore no controls shown.",
      },
    },
  },
};

export const OverriddenContentStyles = {
  render: () => (
    <Expander
      className="bg-neutral-400 p-16 rounded-lg"
      fadeClassName="from-neutral-800"
    >
      {longContentInner}
    </Expander>
  ),
  parameters: {
    docs: {
      description: {
        story:
          "A larger amount of content, with overridden styles for the content wrapper and fader.",
      },
    },
  },
};

export const OverriddenControls = {
  render: () => (
    <Expander
      controlsClassName="ui-btn text-white w-full ui-text-p1 border rounded-xl hover:text-white"
      controlsOpenedLabel={
        <span className="flex items-center gap-8">
          Away with you, knave.{" "}
          <Icon color="text-pink-500" size="24" name="icon-gui-warning" />
        </span>
      }
      controlsClosedLabel="Give me more!"
    >
      {longContentInner}
    </Expander>
  ),
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await expect(canvas.getByTestId("expander-container")).toHaveStyle({
      height: "200px",
    });

    await waitFor(() =>
      expect(canvas.getByTestId("expander-controls")).toBeInTheDocument(),
    );

    await userEvent.click(canvas.getByTestId("expander-controls"));

    await expect(canvas.getByTestId("expander-controls")).toHaveTextContent(
      "Away with you, knave.",
    );

    await waitFor(() =>
      expect(canvas.getByTestId("expander-container")).toHaveStyle({
        height: "664px",
      }),
    );
  },
  parameters: {
    docs: {
      description: {
        story:
          "An expander with overridden styles and labels for the controls.",
      },
    },
  },
};
