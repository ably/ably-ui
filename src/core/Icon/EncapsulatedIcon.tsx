import React from "react";
import Icon, { IconProps } from "../Icon";
import { IconSize } from "./types";

type EncapsulatedIconProps = {
  className?: string;
  innerClassName?: string;
  iconSize?: IconSize;
} & IconProps;

const ICON_SIZE_REDUCTION = 12;
const ICON_HEIGHT_REDUCTION = 2;

const EncapsulatedIcon = ({
  size = "40px",
  iconSize,
  className,
  innerClassName,
  ...iconProps
}: EncapsulatedIconProps) => {
  let computedIconSize = size,
    computedIconHeight = size;

  if (iconSize) {
    computedIconSize = iconSize;
  } else if (size.toString().endsWith("px")) {
    const numericalSize = parseInt(size, 10);
    computedIconSize = `${numericalSize - ICON_SIZE_REDUCTION}px`;
    computedIconHeight = `${numericalSize - ICON_HEIGHT_REDUCTION}px`;
  } else {
    console.warn(
      `EncapsulatedIcon: Non-pixel units might not behave as expected`,
    );
  }

  return (
    <div
      className={`p-1 rounded-lg bg-gradient-to-t dark:bg-gradient-to-b ${className ?? "from-neutral-400 dark:from-neutral-900"}`}
      style={{ width: size, height: size }}
    >
      <div
        className={`flex items-center justify-center rounded-lg ${innerClassName ?? "bg-neutral-200 dark:bg-neutral-1100"}`}
        style={{ height: computedIconHeight }}
      >
        <Icon size={computedIconSize} {...iconProps} />
      </div>
    </div>
  );
};

export default EncapsulatedIcon;
