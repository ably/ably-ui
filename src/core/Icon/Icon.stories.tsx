import React from "react";
import Icon from "../Icon";
import EncapsulatedIcon from "./EncapsulatedIcon";
import loadIcons from "../icons";
import { IconName, iconNames } from "./types";

export default {
  title: "JS Components/Icon",
  component: Icon,
  parameters: {
    layout: "fullscreen",
  },
  tags: ["autodocs"],
};

const renderIcons = (iconSet: IconName[], encapsulated?: boolean) => {
  loadIcons();

  return (
    <div className="grid grid-cols-2 sm:grid-cols-4 ui-grid-gap max-w-screen-lg mb-64">
      {iconSet.map((icon) => (
        <div
          className="border p-16 flex flex-col sm:flex-row items-center justify-between gap-12 rounded-lg"
          key={icon}
        >
          <code className="ui-text-code text-neutral-1300 dark:text-neutral-000 mb-8 block">
            {icon}
          </code>
          <div className="border inline-flex flex-0">
            <div className="flex pi-checkered-bg">
              {encapsulated ? (
                <EncapsulatedIcon name={icon} />
              ) : (
                <Icon
                  name={icon}
                  additionalCSS="hover:text-active-orange"
                  color="text-cool-black"
                  size="1.5rem"
                />
              )}
            </div>
          </div>
        </div>
      ))}
    </div>
  );
};

export const GUIIcons = {
  render: () => renderIcons([...iconNames.gui]),
};

export const DisplayIcons = {
  render: () => renderIcons([...iconNames.display]),
};

export const SocialIcons = {
  render: () => renderIcons([...iconNames.social]),
};

export const TechIcons = {
  render: () => renderIcons([...iconNames.tech]),
};

export const ProductIcons = {
  render: () => renderIcons([...iconNames.product]),
};

export const EncapsulatedIcons = {
  render: () => renderIcons([...iconNames.product], true),
};

export const OtherIcons = {
  render: () => renderIcons([...iconNames.other]),
};

export const IconWithSecondaryColor = {
  render: () => (
    <div className="flex items-center justify-center p-16 gap-24">
      <div className="flex gap-24">
        <h4 className="ui-text-h4 text-neutral-1300 dark:text-neutral-000">
          Non-themed:
        </h4>
        <Icon
          name="icon-gui-check-circled-fill"
          color="text-orange-600"
          size="1.5rem"
        />
        <Icon
          name="icon-gui-check-circled-fill"
          color="text-orange-600"
          secondaryColor="text-neutral-1300"
          size="1.5rem"
        />
      </div>
      <div className="flex gap-24">
        <h4 className="ui-text-h4 text-neutral-1300 dark:text-neutral-000">
          Themed:
        </h4>
        <Icon
          name="icon-gui-check-circled-fill"
          color="text-orange-600"
          size="1.5rem"
        />
        <Icon
          name="icon-gui-check-circled-fill"
          color="text-orange-200 dark:text-orange-1000"
          secondaryColor="text-orange-600 dark:text-orange-600"
          size="1.5rem"
        />
      </div>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story:
          "By setting the `secondaryColor` prop, you can change the color of the icon's secondary elements, if supported in the underlying SVG. You can set either a CSS variable directly, or use a Tailwind class.\n\nThe right-hand icon here (`icon-gui-check-circled-fill`) has this property applied, resulting in it having a dark checkmark in the non-themed case, and inverted colors in the themed case.\n\nTo add support to an existing SVG, replace a `fill` attribute value with `var(--ui-icon-secondary-color)`.",
      },
    },
  },
};
