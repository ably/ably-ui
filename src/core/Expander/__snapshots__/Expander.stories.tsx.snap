// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS Components/Expander LongContent smoke-test 1`] = `
<div data-testid="expander-container"
     class="overflow-hidden transition-all relative "
     style="height: 200px;"
>
  <div class="h-64 w-full bg-gradient-to-t from-white to-transparent absolute bottom-0 left-0 right-0 ">
  </div>
  <div>
    <div>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </li>
        <li>
          Sed convallis ex pharetra, tristique tellus vel, rhoncus velit.
        </li>
        <li>
          Mauris molestie felis et scelerisque ullamcorper.
        </li>
        <li>
          Maecenas congue ligula ut commodo tristique.
        </li>
        <li>
          Pellentesque venenatis elit vitae urna condimentum, in mollis arcu venenatis.
        </li>
        <li>
          Donec nec turpis vel urna egestas fringilla.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Mauris ut nibh vel metus cursus semper.
        </li>
        <li>
          Ut mattis tortor eu urna accumsan gravida.
        </li>
        <li>
          Nunc pellentesque neque at elit pretium tempor.
        </li>
        <li>
          Curabitur finibus magna vitae nunc varius fermentum.
        </li>
      </ul>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Curabitur vehicula mi iaculis, luctus augue eu, venenatis quam.
        </li>
        <li>
          Praesent in eros efficitur, consequat ante eu, faucibus arcu.
        </li>
        <li>
          Nulla laoreet nibh a odio interdum, non molestie diam auctor.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Praesent aliquam diam tincidunt, sollicitudin tortor eget, vulputate lacus.
        </li>
        <li>
          Quisque in mi sed ex vulputate varius in a leo.
        </li>
        <li>
          Etiam posuere dolor at tortor aliquam imperdiet.
        </li>
        <li>
          Maecenas quis neque consequat, ultricies est sit amet, congue est.
        </li>
        <li>
          Aenean a elit sed nibh pretium lacinia sed convallis sapien.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Nulla malesuada libero id dolor aliquam, non sagittis mi scelerisque.
        </li>
        <li>
          Etiam tincidunt lacus eu diam laoreet consectetur sit amet non est.
        </li>
        <li>
          In porta arcu nec purus tincidunt vulputate.
        </li>
      </ul>
    </div>
  </div>
</div>
<div tabindex="0"
     data-testid="expander-controls"
     class="mt-16 cursor-pointer font-bold text-gui-blue-default-light hover:text-gui-blue-hover-light "
>
  View all +
</div>
`;

exports[`JS Components/Expander OverriddenContentStyles smoke-test 1`] = `
<div data-testid="expander-container"
     class="overflow-hidden transition-all relative bg-neutral-400 p-16 rounded-lg"
     style="height: 200px;"
>
  <div class="h-64 w-full bg-gradient-to-t from-white to-transparent absolute bottom-0 left-0 right-0 from-neutral-800">
  </div>
  <div>
    <div>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </li>
        <li>
          Sed convallis ex pharetra, tristique tellus vel, rhoncus velit.
        </li>
        <li>
          Mauris molestie felis et scelerisque ullamcorper.
        </li>
        <li>
          Maecenas congue ligula ut commodo tristique.
        </li>
        <li>
          Pellentesque venenatis elit vitae urna condimentum, in mollis arcu venenatis.
        </li>
        <li>
          Donec nec turpis vel urna egestas fringilla.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Mauris ut nibh vel metus cursus semper.
        </li>
        <li>
          Ut mattis tortor eu urna accumsan gravida.
        </li>
        <li>
          Nunc pellentesque neque at elit pretium tempor.
        </li>
        <li>
          Curabitur finibus magna vitae nunc varius fermentum.
        </li>
      </ul>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Curabitur vehicula mi iaculis, luctus augue eu, venenatis quam.
        </li>
        <li>
          Praesent in eros efficitur, consequat ante eu, faucibus arcu.
        </li>
        <li>
          Nulla laoreet nibh a odio interdum, non molestie diam auctor.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Praesent aliquam diam tincidunt, sollicitudin tortor eget, vulputate lacus.
        </li>
        <li>
          Quisque in mi sed ex vulputate varius in a leo.
        </li>
        <li>
          Etiam posuere dolor at tortor aliquam imperdiet.
        </li>
        <li>
          Maecenas quis neque consequat, ultricies est sit amet, congue est.
        </li>
        <li>
          Aenean a elit sed nibh pretium lacinia sed convallis sapien.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Nulla malesuada libero id dolor aliquam, non sagittis mi scelerisque.
        </li>
        <li>
          Etiam tincidunt lacus eu diam laoreet consectetur sit amet non est.
        </li>
        <li>
          In porta arcu nec purus tincidunt vulputate.
        </li>
      </ul>
    </div>
  </div>
</div>
<div tabindex="0"
     data-testid="expander-controls"
     class="mt-16 cursor-pointer font-bold text-gui-blue-default-light hover:text-gui-blue-hover-light "
>
  View all +
</div>
`;

exports[`JS Components/Expander OverriddenControls play-test 1`] = `
<div data-testid="expander-container"
     class="overflow-hidden transition-all relative "
     style="height: 664px;"
>
  <div>
    <div>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </li>
        <li>
          Sed convallis ex pharetra, tristique tellus vel, rhoncus velit.
        </li>
        <li>
          Mauris molestie felis et scelerisque ullamcorper.
        </li>
        <li>
          Maecenas congue ligula ut commodo tristique.
        </li>
        <li>
          Pellentesque venenatis elit vitae urna condimentum, in mollis arcu venenatis.
        </li>
        <li>
          Donec nec turpis vel urna egestas fringilla.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Mauris ut nibh vel metus cursus semper.
        </li>
        <li>
          Ut mattis tortor eu urna accumsan gravida.
        </li>
        <li>
          Nunc pellentesque neque at elit pretium tempor.
        </li>
        <li>
          Curabitur finibus magna vitae nunc varius fermentum.
        </li>
      </ul>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Curabitur vehicula mi iaculis, luctus augue eu, venenatis quam.
        </li>
        <li>
          Praesent in eros efficitur, consequat ante eu, faucibus arcu.
        </li>
        <li>
          Nulla laoreet nibh a odio interdum, non molestie diam auctor.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Praesent aliquam diam tincidunt, sollicitudin tortor eget, vulputate lacus.
        </li>
        <li>
          Quisque in mi sed ex vulputate varius in a leo.
        </li>
        <li>
          Etiam posuere dolor at tortor aliquam imperdiet.
        </li>
        <li>
          Maecenas quis neque consequat, ultricies est sit amet, congue est.
        </li>
        <li>
          Aenean a elit sed nibh pretium lacinia sed convallis sapien.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Nulla malesuada libero id dolor aliquam, non sagittis mi scelerisque.
        </li>
        <li>
          Etiam tincidunt lacus eu diam laoreet consectetur sit amet non est.
        </li>
        <li>
          In porta arcu nec purus tincidunt vulputate.
        </li>
      </ul>
    </div>
  </div>
</div>
<div tabindex="0"
     data-testid="expander-controls"
     class="mt-16 cursor-pointer font-bold text-gui-blue-default-light hover:text-gui-blue-hover-light ui-btn text-white w-full ui-text-p1 border rounded-xl hover:text-white"
>
  <span class="flex items-center gap-8">
    Away with you, knave.
    <svg class="text-pink-500"
         style="width: 24px; height: 24px;"
    >
      <use xlink:href="#sprite-icon-gui-warning">
      </use>
    </svg>
  </span>
</div>
`;

exports[`JS Components/Expander ShortContent smoke-test 1`] = `
<div data-testid="expander-container"
     class="overflow-hidden transition-all relative "
     style="height: auto;"
>
  <div>
    <div>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </li>
        <li>
          Sed convallis ex pharetra, tristique tellus vel, rhoncus velit.
        </li>
        <li>
          Mauris molestie felis et scelerisque ullamcorper.
        </li>
        <li>
          Maecenas congue ligula ut commodo tristique.
        </li>
        <li>
          Pellentesque venenatis elit vitae urna condimentum, in mollis arcu venenatis.
        </li>
        <li>
          Donec nec turpis vel urna egestas fringilla.
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`JS Components/Expander ZeroHeightContent smoke-test 1`] = `
<div data-testid="expander-container"
     class="overflow-hidden transition-all relative "
     style="height: 0px;"
>
  <div class="h-64 w-full bg-gradient-to-t from-white to-transparent absolute bottom-0 left-0 right-0 from-transparent">
  </div>
  <div>
    <div>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </li>
        <li>
          Sed convallis ex pharetra, tristique tellus vel, rhoncus velit.
        </li>
        <li>
          Mauris molestie felis et scelerisque ullamcorper.
        </li>
        <li>
          Maecenas congue ligula ut commodo tristique.
        </li>
        <li>
          Pellentesque venenatis elit vitae urna condimentum, in mollis arcu venenatis.
        </li>
        <li>
          Donec nec turpis vel urna egestas fringilla.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Mauris ut nibh vel metus cursus semper.
        </li>
        <li>
          Ut mattis tortor eu urna accumsan gravida.
        </li>
        <li>
          Nunc pellentesque neque at elit pretium tempor.
        </li>
        <li>
          Curabitur finibus magna vitae nunc varius fermentum.
        </li>
      </ul>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Curabitur vehicula mi iaculis, luctus augue eu, venenatis quam.
        </li>
        <li>
          Praesent in eros efficitur, consequat ante eu, faucibus arcu.
        </li>
        <li>
          Nulla laoreet nibh a odio interdum, non molestie diam auctor.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Praesent aliquam diam tincidunt, sollicitudin tortor eget, vulputate lacus.
        </li>
        <li>
          Quisque in mi sed ex vulputate varius in a leo.
        </li>
        <li>
          Etiam posuere dolor at tortor aliquam imperdiet.
        </li>
        <li>
          Maecenas quis neque consequat, ultricies est sit amet, congue est.
        </li>
        <li>
          Aenean a elit sed nibh pretium lacinia sed convallis sapien.
        </li>
      </ul>
      <p>
        Ipsum
      </p>
      <ul class="mb-16 list-inside list-disc">
        <li>
          Nulla malesuada libero id dolor aliquam, non sagittis mi scelerisque.
        </li>
        <li>
          Etiam tincidunt lacus eu diam laoreet consectetur sit amet non est.
        </li>
        <li>
          In porta arcu nec purus tincidunt vulputate.
        </li>
      </ul>
    </div>
  </div>
</div>
<div tabindex="0"
     data-testid="expander-controls"
     class=" cursor-pointer font-bold text-gui-blue-default-light hover:text-gui-blue-hover-light "
>
  View all +
</div>
`;
